pkg/template/api/types.go

```diff
@@ -75,3 +75,137 @@ type Parameter struct {
+// TemplateInstance is part of an experimental API.
+type TemplateInstance struct {
+	unversioned.TypeMeta
+	kapi.ObjectMeta
+
+	// Spec describes the desired state of this TemplateInstance.
+	Spec TemplateInstanceSpec
+
+	// Status describes the current state of this TemplateInstance.
+	Status TemplateInstanceStatus
+}
+
+// TemplateInstanceSpec describes the desired state of a TemplateInstance.
+type TemplateInstanceSpec struct {
+	// Template is a full copy of the template for instantiation.
+	Template Template
+
+	// SecretRef is a reference to a Secret object containing the necessary
+	// template parameters.
+	SecretRef kapi.LocalObjectReference
+
+	// Requestor holds the identity of the agent requesting the template
+	// instantiation.
+	Requestor *TemplateInstanceRequestor
+}
+
+// TemplateInstanceRequestor holds the identity of an agent requesting a
+// template instantiation.
+type TemplateInstanceRequestor struct {
+	// Username is the username of the agent requesting a template instantiation.
+	Username string
+}
+
+// TemplateInstanceStatus describes the current state of a TemplateInstance.
+type TemplateInstanceStatus struct {
+	// Conditions represent the latest available observations of a
+	// TemplateInstance's current state.
+	Conditions []TemplateInstanceCondition
+}
+
+// TemplateInstanceCondition contains condition information for a
+// TemplateInstance.
+type TemplateInstanceCondition struct {
+	// Type of the condition, currently Ready or InstantiateFailure.
+	Type TemplateInstanceConditionType
+	// Status of the condition, one of True, False or Unknown.
+	Status kapi.ConditionStatus
+	// LastTransitionTime is the last time a condition status transitioned from
+	// one state to another.
+	LastTransitionTime unversioned.Time
+	// Reason is a brief machine readable explanation for the condition's last
+	// transition.
+	Reason string
+	// Message is a human readable description of the details of the last
+	// transition, complementing reason.
+	Message string
+}
+
+// TemplateInstanceConditionType is the type of condition pertaining to a
+// TemplateInstance.
+type TemplateInstanceConditionType string
+
+const (
+	// TemplateInstanceReady indicates the readiness of the template
+	// instantiation.
+	TemplateInstanceReady TemplateInstanceConditionType = "Ready"
+	// TemplateInstanceInstantiateFailure indicates the failure of the template
+	// instantiation
+	TemplateInstanceInstantiateFailure TemplateInstanceConditionType = "InstantiateFailure"
+)
+
+// TemplateInstanceList is a list of TemplateInstance objects.
+type TemplateInstanceList struct {
+	unversioned.TypeMeta
+	unversioned.ListMeta
+
+	// Items is a list of Templateinstances
+	Items []TemplateInstance
+}
+
+// TemplateBrokerUUIDMap holds the service broker-related state associated with
+// a TemplateInstance.  TemplateBrokerUUIDMap is part of an experimental API.
+type TemplateBrokerUUIDMap struct {
```

Comment:
@smarterclayton told me I wasn't living up to my reputation for being picky about naming, so...  given that UUID is an implementation detail, perhaps this should be named differently.

TemplateBrokerProvisionMap?  TemplateInstanceMap?  TemplateInstanceProvisionMap? TemplateBrokerProvisionInstanceSpec?  TemplateInstanceProvisionSpec? 

(I don't know that it really needs "map" in the name)